std/hash/auto_hash.zig

```diff
@@ -0,0 +1,208 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const mem = std.mem;
+const meta = std.meta;
+
+/// Provides generic hashing for any eligible type.
+/// Only hashes `key` itself, pointers are not followed.
+pub fn autoHash(hasher: var, key: var) void {
+    const Key = @typeOf(key);
+    switch (@typeInfo(Key)) {
+        builtin.TypeId.NoReturn,
+        builtin.TypeId.Opaque,
+        builtin.TypeId.Undefined,
+        builtin.TypeId.ArgTuple,
+        builtin.TypeId.Void,
+        builtin.TypeId.Null,
+        builtin.TypeId.BoundFn,
+        builtin.TypeId.ComptimeFloat,
+        builtin.TypeId.ComptimeInt,
+        builtin.TypeId.Type,
+        builtin.TypeId.EnumLiteral,
+        => @compileError("cannot hash this type"),
+
+        builtin.TypeId.Int => hasher.update(std.mem.asBytes(&key)),
+
+        builtin.TypeId.Float => |info| autoHash(hasher, @bitCast(@IntType(false, info.bits), key)),
+
+        builtin.TypeId.Bool => autoHash(hasher, @boolToInt(key)),
+        builtin.TypeId.Enum => autoHash(hasher, @enumToInt(key)),
+        builtin.TypeId.ErrorSet => autoHash(hasher, @errorToInt(key)),
+        builtin.TypeId.Promise, builtin.TypeId.Fn => autoHash(hasher, @ptrToInt(key)),
+
+        builtin.TypeId.Pointer => |info| switch (info.size) {
+            builtin.TypeInfo.Pointer.Size.One,
+            builtin.TypeInfo.Pointer.Size.Many,
+            builtin.TypeInfo.Pointer.Size.C,
+            => autoHash(hasher, @ptrToInt(key)),
+
+            builtin.TypeInfo.Pointer.Size.Slice => {
+                autoHash(hasher, key.ptr);
+                autoHash(hasher, key.len);
+            },
+        },
+
+        builtin.TypeId.Optional => if (key) |k| autoHash(hasher, k),
+
+        builtin.TypeId.Array => {
+            // TODO detect via a trait when Key has no padding bits to
+            // hash it as an array of bytes.
+            // Otherwise, hash every element.
+            for (key) |element| {
+                autoHash(hasher, element);
+            }
+        },
+
+        builtin.TypeId.Vector => |info| {
+            if (info.child.bit_count % 8 == 0) {
+                // If there's no unused bits in the child type, we can just hash
+                // this as an array of bytes.
+                hasher.update(mem.asBytes(&key));
+            } else {
+                // Otherwise, hash every element.
+                // TODO remove the copy to an array once field access is done.
+                const array: [info.len]info.child = key;
+                comptime var i: u32 = 0;
+                inline while (i < info.len) : (i += 1) {
+                    autoHash(hasher, array[i]);
+                }
+            }
+        },
+
+        builtin.TypeId.Struct => |info| {
+            // TODO detect via a trait when Key has no padding bits to
+            // hash it as an array of bytes.
+            // Otherwise, hash every field.
```

Comment:
For now you could at least hash the whole thing when `ContainerLayout` is `Packed`.